mofa/model.py

```diff
@@ -28,38 +36,144 @@ class NodeDescription:
+
+    def create_description(self, atom_types: list[str], coordinates: np.ndarray) -> 'LigandDescription':
+        """Produce a ligand description given atomic coordinates which include the infilled atoms
+
+        Args:
+            atom_types: Types of all atoms as chemical symbols
+            coordinates: Coordinates of all atoms
+        Returns:
+            Ligand description using the new coordinates
+        """
+
+        # The linker groups should be up from, make sure the types have not changed and change if they have
+        pos = 0
+        anchor_atoms = []
+        for anchor in self.anchors:
+            # Determine the fragment positions
+            found_types = atom_types[pos:pos + len(anchor)]
+            anchor_atoms.append(list(range(pos, pos + len(anchor))))
+            expected_types = anchor.get_chemical_symbols()
+
+            # Make sure the types match, and increment position
+            assert found_types == expected_types, f'Anchor changed. Found: {found_types} - Expected: {expected_types}'
+            pos += len(anchor)
+
+        # Build the XYZ file
+        atoms = ase.Atoms(symbols=atom_types, positions=coordinates)
+        return LigandDescription(
+            xyz=write_to_string(atoms, 'xyz'),
+            role=self.role,
+            anchor_atoms=anchor_atoms,
+            dummy_element=self.dummy_element
+        )
+
+    @classmethod
+    def from_yaml(cls, path: Path) -> 'LigandTemplate':
+        """Load a template description from YAML
+
+        Args:
+            path: Path to the YAML file
+        Returns:
+            The ligand template
+        """
+
+        with path.open() as fp:
+            return cls(**yaml.safe_load(fp))
+
+
 @dataclass
 class LigandDescription:
     """Description of organic sections which connect inorganic nodes"""
 
-    smiles: str = field()
+    smiles: str | None = field(default=None)
     """SMILES-format designation of the molecule"""
     xyz: str | None = field(default=None, repr=False)
     """XYZ coordinates of each atom in the linker"""
     role: str | None = field(default=None)
     """Portion of the MOF to which this ligand corresponds"""
 
-    fragment_atoms: list[list[int]] | None = field(default=None, repr=True)
+    anchor_atoms: list[list[int]] | None = field(default=None, repr=True)
     """Groups of atoms which attach to the nodes
 
     There are typically two groups of fragment atoms, and these are
     never altered during MOF generation."""
+    dummy_element: str = field(default=None, repr=False)
+    """Element used to represent the end group during assembly"""
 
-    @property
-    def linker_atoms(self) -> list[int]:
-        """All atoms which are not part of a fragment"""
-        raise NotImplementedError()
+    @cached_property
+    def atoms(self):
+        return read_from_string(self.xyz, "xyz")
+
+    def replace_with_dummy_atoms(self, anchor_type: str = "COO") -> ase.Atoms:
```

Comment:
Should we store the anchor type as part of the template name? That way we don't have to pass it as an input to this function?

Reply:
I went ahead and made this change